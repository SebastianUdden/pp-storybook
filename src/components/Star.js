import React from "react";
import styled from "styled-components";

const Svg = styled.svg`
  transform: rotate(55deg);
  width: ${p => p.width};
  cursor: ${p => (p.fixedAmount ? "default" : "pointer")};
  :hover {
    opacity: ${p => (p.fixedAmount ? 1 : 0.5)};
  }
`;

const calcStarPoints = (
  centerX,
  centerY,
  innerCirclePoints,
  innerRadius,
  outerRadius
) => {
  const angle = Math.PI / innerCirclePoints;
  const angleOffsetToCenterStar = 0;

  const totalPoints = innerCirclePoints * 2; // 10 in a 5-points star
  let points = "";

  for (let i = 0; i < totalPoints; i++) {
    let isEvenIndex = i % 2 == 0;
    let r = isEvenIndex ? outerRadius : innerRadius;
    let currX = centerX + Math.cos(i * angle + angleOffsetToCenterStar) * r;
    let currY = centerY + Math.sin(i * angle + angleOffsetToCenterStar) * r;
    points += currX + "," + currY + " ";
  }
  return points;
};

const getAdjustedPercentage = (high, low, ratio) => (high - low) * ratio + low;

const Star = ({
  id,
  color,
  remainderRatio,
  width,
  height,
  onClick,
  fixedAmount
}) => {
  const fill = getAdjustedPercentage(70, 30, remainderRatio);
  const getStarPoints = (width, height) => {
    let centerX = width / 2; // e.g 100
    let centerY = height / 2; // e.g 100

    let innerCirclePoints = 5; // a 5 point star

    // width --> this is the beam length of each
    // side of the SVG square that holds the star
    let innerRadius = width / innerCirclePoints;
    let innerOuterRadiusRatio = 2.5; // outter circle is x2 the inner
    let outerRadius = innerRadius * innerOuterRadiusRatio;

    return calcStarPoints(
      centerX,
      centerY,
      innerCirclePoints,
      innerRadius,
      outerRadius
    );
  };

  return (
    <Svg
      viewBox="0 0 24 24"
      width={width}
      onClick={onClick}
      fixedAmount={fixedAmount}
    >
      <defs>
        <linearGradient
          id={id}
          gradientTransform="rotate(-55)"
          x1="-130%"
          y1="0"
        >
          <stop stop-color={color} offset={`${fill}%`} />
          <stop stop-color="grey" offset="0%" />
          <stop stop-color="grey" stop-opacity="0" offset="0%" />
        </linearGradient>
      </defs>
      <polygon
        points={getStarPoints(width, height)}
        fill={`url(#${id})`}
        style={{ fillRule: "nonzero" }}
      />
    </Svg>
  );
};

export default Star;
