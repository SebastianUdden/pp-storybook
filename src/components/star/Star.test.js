import React from "react";
import renderer from "react-test-renderer";
import Star from "./Star";

export const actions = {};
export const attributes = {
  width: 24,
  height: 24
};

test("Star renders correctly", () => {
  const tree = renderer
    .create(
      <Star
        id="Default"
        color={"orange"}
        remainderRatio={0}
        {...attributes}
        {...actions}
        onClick={() => setColor("grey")}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
