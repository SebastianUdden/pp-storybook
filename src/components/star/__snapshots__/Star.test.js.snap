// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Star renders correctly 1`] = `
<svg
  className="sc-bdVaJa iRnrFu"
  onClick={[Function]}
  viewBox="0 0 24 24"
  width={24}
>
  <defs>
    <linearGradient
      gradientTransform="rotate(-55)"
      id="Default"
      x1="-134%"
      y1="0"
    >
      <stop
        offset="30%"
        stop-color="orange"
      />
      <stop
        offset="0%"
        stop-color="grey"
      />
      <stop
        offset="0%"
        stop-color="grey"
        stop-opacity="0"
      />
    </linearGradient>
  </defs>
  <polygon
    fill="url(#Default)"
    points="24,12 15.883281572999747,14.821369211003871 15.70820393249937,23.412678195541844 10.516718427000253,16.565071278216738 2.2917960675006324,19.05342302750968 7.2,12 2.2917960675006306,4.946576972490323 10.516718427000251,7.434928721783264 15.708203932499366,0.5873218044581563 15.883281572999747,9.178630788996127 "
    style={
      Object {
        "fillRule": "nonzero",
      }
    }
  />
</svg>
`;
